metals/src/main/scala/scala/meta/internal/metals/InteractiveSemanticdbs.scala

```diff
@@ -170,34 +162,79 @@ final class InteractiveSemanticdbs(
       }
       .getOrElse(compilers().ramboCompiler)
 
+    val (prependedLinesSize, modifiedText) =
+      buildTargets
+        .sbtAutoImports(source)
+        .fold((0, text))(imports =>
+          (imports.size, SbtBuildTool.prependAutoImports(text, imports))
+        )
+
     val uri = source.toURI.toString
     // NOTE(olafur): it's unfortunate that we block on `semanticdbTextDocument`
     // here but to avoid it we would need to refactor the `Semanticdbs` trait,
     // which requires more effort than it's worth.
     val bytes = pc
-      .semanticdbTextDocument(uri, text)
+      .semanticdbTextDocument(uri, modifiedText)
       .get(
         clientConfig.initialConfig.compilers.timeoutDelay,
         clientConfig.initialConfig.compilers.timeoutUnit
       )
-    val textDocument = TextDocument.parseFrom(bytes)
-    textDocument
-
+    val textDocument = s.TextDocument.parseFrom(bytes)
+    if (prependedLinesSize > 0)
+      cleanupAutoImports(textDocument, text, prependedLinesSize)
```

Comment:
In the beginning, I thought to implement it as you suggest.

However, I found that `references` and `rename` are based on `DefinitionProvider.positionOccurence`.
It internally uses `TokenEditDistance` between buffer and TextDocument. So for sbt files, it means that TextDocument has to be cleaned from sbt-specific imports before it will be passed into `TokenEditDistance`. Otherwise, it's method `toOriginal` fails.
It might be handled by adding sbt/adjust-specific logic into `DefinitionProvider` but I don't think that it's the right way to go.

So, I decided that it would be better just to leave these hacks in `IntercativeSemanticdb` because in this case adjustment is applied on TextDocument, not on LSP positions.

Reply:
I think that should not be an issue. DefinitionProvider would still use `InteractiveSemanticdb`, so there would be no need to add sbt specific logic anywhere. However, I see now that it's a bigger refactor here, so maybe I can take a look a bit later while taking a look at how we infer build targets (especially for worksheets, which I guess might not work perfectly now) and how we could improve it.